### PR TITLE
Ns1 provider

### PR DESCRIPTION
**Pull request by stack72**
This is a port of #10172 

We have renamed nsone to ns1 (as per a conversation with @josephholsten)

### CODE REVIEW

**Comment by stack72** [comment]
@josephholsten this is now ready for the PR that we spoke about (meta??) to be raised against. Please also be aware that this is currently missing documentation. We only have 2 resources documented

**Comment by josephholsten** [comment]
@pashap This is relevant to your interests.

**Comment by pashap** [comment]
@josephholsten Thanks, did you guys get a chance to check out https://github.com/hashicorp/terraform/pull/10391 ? (I understand if not, there are so many branches/PRs for this provider :) )

Its an extended branch with work that @cakoose did on meta and had all the tests passing except for `TestAccRecord` bc of our ns1 metadata issue.

File: builtin/providers/ns1/permissions.go [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+func addPermsSchema(s map[string]*schema.Schema) map[string]*schema.Schema {
+	s["dns_view_zones"] = &schema.Schema{
+		Type:     schema.TypeBool,
+		Optional: true,
+	}
+	s["dns_manage_zones"] = &schema.Schema{
+		Type:     schema.TypeBool,
+		Optional: true,
+	}
+	s["dns_zones_allow_by_default"] = &schema.Schema{
```
Comment (pashap) [inline builtin/providers/ns1/permissions.go]: Hi @jbardin This comment is related to a failing acceptance test:

```
 $ make testacc TEST=./builtin/providers/ns1 TESTARGS='-run=TestAccTeam_updated'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/01/20 13:48:12 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/ns1 -v -run=TestAccTeam_updated -timeout 120m
=== RUN   TestAccTeam_updated
panic: empty structure "dns_zones_allow" has entry "dns_zones_allow_by_default"

goroutine 430 [running]:
panic(0x51e340, 0xc42048e2f0)
        /usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/hashicorp/terraform/terraform.(*InstanceState).MergeDiff(0xc4202baeb0, 0xc4202d83e0, 0x0)
```

I could very easily be reading this wrong, but I believe that https://github.com/hashicorp/terraform/commit/2820845f8c3b21cb35666c29fa276da535ab99dd L1672 is supposed to check that the attribute is of either an array, map, or set?

The schema at fault in the ns1 provider has some subfields that share the same prefix(`dns_zones`):
 - `dns_zones_allow_by_default` as `schema.TypeBool`
 - `dns_zones_allow` as `schema.TypeList`
 - `dns_zones_deny` as `schema.TypeList`


I believe that the boolean `dns_zones_allow_by_default` shouldnt be included in that sanity check, am i thinking correctly on this?

Comment (jbardin) [inline builtin/providers/ns1/permissions.go]: I think I know here the bug is here. I have another repro case that needs fixing. I'll et a patch into master ASAP and see if it fixing your issue here.

Comment (pashap) [inline builtin/providers/ns1/permissions.go]: Thank you sir!

Comment (jbardin) [inline builtin/providers/ns1/permissions.go]: Should be fixed on master now if you want to rebase.

You should probably drop your vendor commit, and re-add those packages correctly using govendor after rebasing, as that will fail a check now too.

Comment (pashap) [inline builtin/providers/ns1/permissions.go]: Sweet, will do. Thanks again. Had the govendor fix stuff in a different branch(https://github.com/hashicorp/terraform/pull/10391/commits/3dfcfc21265628c22d843f8ea99b51321f50e9cc)... this one just got out of hand

**Comment by stack72** [comment]
@pashap @josephholsten this now LGTM! Thanks for all the work here :)

```
% make testacc TEST=./builtin/providers/ns1                                                                                               ✭
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/01/23 23:28:40 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/ns1 -v  -timeout 120m
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccDataFeed_basic
--- PASS: TestAccDataFeed_basic (13.83s)
=== RUN   TestAccDataFeed_updated
--- PASS: TestAccDataFeed_updated (11.69s)
=== RUN   TestAccDataSource_basic
--- PASS: TestAccDataSource_basic (6.95s)
=== RUN   TestAccDataSource_updated
--- PASS: TestAccDataSource_updated (15.32s)
=== RUN   TestAccMonitoringJob_basic
--- PASS: TestAccMonitoringJob_basic (6.30s)
=== RUN   TestAccMonitoringJob_updated
--- PASS: TestAccMonitoringJob_updated (11.09s)
=== RUN   TestAccRecord_basic
--- PASS: TestAccRecord_basic (13.67s)
=== RUN   TestAccRecord_updated
--- PASS: TestAccRecord_updated (20.22s)
=== RUN   TestAccTeam_basic
--- PASS: TestAccTeam_basic (6.12s)
=== RUN   TestAccTeam_updated
--- PASS: TestAccTeam_updated (8.83s)
=== RUN   TestAccZone_basic
--- PASS: TestAccZone_basic (18.28s)
=== RUN   TestAccZone_updated
--- PASS: TestAccZone_updated (10.50s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/ns1	142.836s
```

**Comment by pashap** [comment]
Awesome! Thanks @stack72 / @josephholsten / @cakoose

Im going to spend some time tonight writing up an issue around our struggles with representing NS1 metadata. Will get something out regarding https://github.com/hashicorp/terraform/pull/11300#issuecomment-274224548

**Pull request merged**